resources/helm/dask-gateway/templates/gateway/configmap.yaml

```
@@ -118,3 +118,8 @@ data:
     {{- end }}
     {{- end }}
     {{- end }}
+
+    {{/* Configuration pulled in through *_config.py files in the extensions directory */}}
+    {{ range $path, $_ :=  .Files.Glob  "**_config.py" }}
```

This globs every directory, we'd probably want to limit to `extensions`? Why not drop the `*_config.py` pattern and add a subdirectory of `extensions` only for extra files to add to the `dask-gateway` pod. Something like:

```suggestion
    {{ range $path, $_ :=  .Files.Glob  "extensions/gateway/*.py" }}
```

JupyterHub follows a similar pattern: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/master/jupyterhub/templates/hub/configmap.yaml#L10

---

Thanks for responding so quickly, Jim.

> Hmmm, interesting. IIUC this requires that users have a local copy of the helm chart, rather than using the published one in the helm repository?

This approach does assume a local copy of the Helm chart. 

> 
> I'm a bit wary of inlining multiple files into the `dask_gateway_config.py` - a folder of python files looks a lot like a python module, I'm afraid users may try to import functionality from one file into another, which would fail in practice. Global variables and imports might also collide, leading to issues.
> 
> Instead of inlining the files in the `dask_gateway_config.py` entry in the configmap, we could use `AsConfig` to add configmap entries for each file placing them in the same directory as the `dask_gateway_config.py` file (we'd need to drop the subdirectories support in `extensions`, as only files would work). You could then have the bulk of your code defining new classes or functions in files in `extensions`, and use `extraConfig` to import the bits you need to pass to dask-gateway's configuration.

Correct me if I'm wrong, but won't this require adding [extraVoleMounts](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/6e5ec7e50a8ed670bae4c5d195bb01a82b157a4b/jupyterhub/templates/hub/deployment.yaml#L122) to the Dask Gateway API deployment and the [values.yaml](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/master/jupyterhub/values.yaml#L55) files?

I want to mention two minor drawbacks. First, it requires updates in two locations: the `extensions` directory and `extraConfig` in `values.yaml`. Second, the extensions cannot be structured like packages, which is less of an issue for `options`, but authentication extensions can get fairly complex. Neither of these are deal breakers, and the alternative, naming collisions, is a significant concern -- I just think they're worth mentioning.

I'm happy to update this PR to align with the JupyterHub design pattern if you think it's a better model. 

> 
> The `AsConfig` helm helper (https://helm.sh/docs/chart_template_guide/accessing_files/#configmap-and-secrets-utility-functions) would help here.

This is a bit of a tangent, and probably deserves a discrete PR, but our authentication extension imports the [pyjwt](https://pyjwt.readthedocs.io/en/latest/) package. I'm wondering if there's a way to side-load that dependency without needing to build our own image or if that's a fool's errand. I suspect other users writing custom authentication modules will run into a similar obstacle.